docs/man/reprounzip.1
.\" Manpage for reprounzip
.TH man 1 "4 November 2017" "1.0.11" "reprounzip man page"
.SH NAME
reprounzip \- the reproducibility unpacker
.SH SYNOPSIS
reprounzip [\-h] [\-\-version] [\-v] ...
.SH DESCRIPTION
reprounzip inspects or unpacks a reproducible experiment package (.rpz file) created by reprozip.
.SH OPTIONS
.SS General options:
.TP
.B \-h, \-\-help
show this help message and exit
.TP
.B \-\-version
show program's version number and exit
.TP
.B \-v, \-\-verbose
augments verbosity level

.SS Commands:
.TP
.B usage_report
Enables or disables anonymous usage reports
.TP
.B info
Prints out some information about a pack
.TP
.BI graph " <output.dot> [package.rpz]"
Generates a provenance graph from the trace data
.TP
.BI showfiles " <package>"
Prints out input and output file names
.TP
.B installpkgs " <package.rpz>"
Installs the required packages on this system
.TP
.BR directory " ..."
Unpacks the files in a directory and runs with PATH and LD_LIBRARY_PATH
.TP
.B chroot
Unpacks the files and run with chroot
.TP
.B docker
Runs the experiment in a Docker container
.TP
.B vagrant
Runs the experiment in a virtual machine created through Vagrant

.SS Unpacker commands:
Each unpacker (such as
.BR directory ", " chroot ", " docker ", or " vagrant )
has a similar set of commands, detailed below. Please refer to
.BR reprounzip\-\f(BIunpacker (1)
for detailed information about a specific unpacker.
.TP
.BI setup " <package.rpz> <directory>"
Unpacks an experiment and sets it up for reproduction in the specified directory.
.TP
.BI run " <directory> [run] [\-\-cmdline ...]"
Runs an experiment that has been unpacked to a directory.
.TP
.BI upload " <directory> [localfile:experimentfile] [...]"
Replaces a file inside the experiment with your own. This allows you to reproduce an experiment in its original environment, but with modified data or configuration. Running this command with only the
.I directory
will list all the named files that you can replace with this command.
.TP
.BI download " <directory> [experimentfile[:localfile]] [...]"
Download a file from the experiment container/virtual machine/directory to your filesystem. If
.I localfile
is not specified, a file of the same name will be written in the current directory. Running this command with only the
.I directory
will list all the named files that you can download with this command.
.TP
.BI destroy " <directory>"
Removes an unpacked experiment directory, and associated resources such as containers, virtual machines, or mount points.
.B Always use this command
to delete an unpacked experiment, otherwise resources might get left behind, or with chroot you might end up
.B deleting system files
through the mount point that has been created.
.SH FILES
.TP
.B ~/.reprozip/log
Log file where reprozip and reprounzip write messages during execution. Useful to report issues to the developers, even if you were running without
.IR \-\-verbose .
.SH ENVIRONMENT
.TP
.B REPROZIP_USAGE_STATS
If this variable is set to
.I \*(lqoff\*(rq
then usage statistics will be completely disabled (won't be reported, recorded, and won't ask you about it).
.TP
.B REPROZIP_PARAMETERS
If this variable is set to a URL, it specifies an alternate location from which to download runtime parameters. If it is set to
.IR \*(lqoff\*(rq ,
nothing will be downloaded and the bundled parameters will be used instead.
.SH EXAMPLES
.P
Get information on a package:
.IP
.nf
.RB "$" " reprounzip \-v info" " foobar0.4\-python2.rpz"
\-\-\-\-\- Pack information \-\-\-\-\-
Compressed size: 1.64 MB
Unpacked size: 4.98 MB
[...]
.fi

.P
Unpack the experiment using
.BR reprounzip\-docker (1)
under
.IR /tmp :
.IP
.nf
.RB "$" " reprounzip docker setup" " foobar0.4\-python2.rpz /tmp/foobar\-docker"
.RB "$" " reprounzip docker run" " /tmp/foobar\-docker"
.RB "$" " reprounzip docker destroy" " /tmp/foobar\-docker"
.fi

.SH SEE ALSO
.TP
The ReproZip website
https://www.reprozip.org/
.TP
ReproZip's GitHub repository
https://github.com/ViDA\-NYU/reprozip
.SH BUGS
Please report bugs on our mailing-list at reprozip@nyu.edu or on GitHub at https://github.com/VIDA\-NYU/reprozip/issues.
.SH AUTHOR
.RB "ReproZip is being developed at" " New York University" .

The team includes:
.RS
.nf
Remi Rampin
Fernando Chirigati
Vicky Rampin
Juliana Freire
Dennis Shasha
.fi
.RE
.SH COPYRIGHT
Copyright 2014 New York University.

.RB "Licensed under a" " BSD 3-Clause license." " See the LICENSE file included with the software for details."


CHANGELOG.md
Changelog
=========

2.0.0 (???)
-----------

Behavior change:
* .rpz config format changed (version 0.8 -> 1.1)

Features:
* Configuration file contains the walltime taken by each run
* It is now possible to upload or download any file via its full path

1.1 (2021-07-06)
----------------

(reprounzip-vistrails didn't change)

Bugfixes:
* Fix possible crash reading `docker inspect` output
* Fix reprozip-jupyter on more recent Tornado
* Fix failure in reprounzip-docker upload if `/bin/sh` is missing
* Have `reprounzip directory` execute with the correct interpreter (e.g. packed `ld-linux.so`)
* Fix "invalid cross-device link" errors in vagrant download if the destination is not on the same device as the unpacked directory

Enhancements:
* Add PyQt5 compatibility to reprounzip-qt
* reprounzip-docker: Keep ownership of uploaded files
* Add a new certificate to download parameters, valid through 2121 (previous expires 2024)
* Always show full path of executed files in `reprozip testrun`
* Improved merging of pack on top of base image in reprounzip-docker
* Made sure the root is a mountpoint in reprounzip-vagrant, as some applications expect it (e.g. Elasticsearch)
* Disable unpacker options that won't work in reprounzip-qt (e.g. 'directory' and 'chroot' on non-Linux)

1.0.16 (2019-02-06)
-------------------

(reprozip-jupyter and reprounzip-vagrant didn't change)

Bugfixes:
* Fixed input/output file filter on Python 3 (to omit `.so`, `.pyc` etc files)
* Fixed fetching updated parameters on Python 3 (to get the correct Docker and Vagrant base images, a small JSON file is downloaded from reprozip.org)
* Fixed `--port` option of reprounzip-docker

Enhancements:
* Use the [distro](https://distro.readthedocs.io/) module instead of the deprecated `platform.linux_distribution()` function to detect the distribution (the latter will be removed in Python 3.8).
* Use dpkg-query to identify Linux packages instead of reading `dpkg/info/*.list`

1.0.15 (2018-07-31)
-------------------

(reprounzip-qt only)

Bugfixes:
* Fixed running command from reprounzip-qt on Windows
* Fixed using Jupyter from reprounzip-qt

1.0.14 (2018-07-30)
-------------------

(reprozip, reprounzip-qt and reprozip-jupyter only)

Bugfixes:
* Fixed reprounzip-qt refusing to close when an experiment is still unpacked, even after the user provided confirmation
* Fixed reprozip-jupyter on Python 3
* Fixed running gnome-terminal from reprounzip-qt, made it preferred over xterm
* Don't duplicate the latest run in the config file when the trace didn't add a run (for example because the command does not exist)

Enhancements:
* Uniformized logos and icons
* Native terminal opened by reprounzip-qt waits for a key after success before closing
* Officially support reprounzip-qt and reprozip-jupyter on Python 3

1.0.13 (2018-05-15)
-------------------

Bugfixes:
* Fix uninitialized return value making some xxx_at() calls abort the trace
* Fix some other warnings via static analysis

Enhancements:
* Show a warning when executing a file that has the set-uid or set-gid bit set, since Linux will not give it its privileges, making it confusing for users why their run failed
* Make reprounzip-docker run even without a TTY
* Correctly handle experiment returning non-0 in Docker
* The C extension now logs through Python's logging facilities ('reprozip' logger)
* Collect usage information from reprounzip-qt as well

1.0.12 (2018-03-30)
-------------------

(reprozip, reprounzip-qt and reprozip-jupyter only)

Bugfixes:
* Fix some kernel/libc issuing unrecognized `openat()` calls, resulting in files missing in the trace
* Fix `openat()` calls recording read-write as simply write
* Fix double-click on .RPZ file

1.0.11 (2017-11-05)
-------------------

Bugfixes:
* Write timestamp in config with timezone offset
* Always fix up PATH on MacOS (to pick up Docker)
* Don't have reprounzip-qt choke on MacOS's `-psn_x_xxxx` arguments
* Fix "Couldn't find reprounzip command" in the GUI on Windows if installed from the installer and opened a package by double-click

Enhancements:
* Improve DOT graph layout
* Don't silently overwrite output with `reprounzip graph`
* Add `--regex-include` to `reprounzip graph`
* `reprozip trace` now returns 125 for tracing errors (previously 1), and returns the traced program's exit code otherwise (previously always 0)
* Manpages are available

1.0.10 (2017-07-10)
-------------------

Bugfixes:
* Correctly escape shell commands containing backticks
* Overwrite tty prompt works correctly on Python 3
* Fix /proc in vagrant-chroot and chroot having outside mounts
* Fix ANSI escapes showing in Qt terminal dialog
* Fix reprozip combine crash on Python 3.6 (patch from James Clarke)
* Using `graph --packages drop` together with `--json` no longer crashes

Enhancements:
* New reprozip-jupyter tool to trace and reproduce Jupyter notebooks
* reprozip_jupyter can be registered as a Jupyter extension to trace notebooks from the Jupyter web interface
* The Qt GUI knows to run packages with reprozip-jupyter if they are notebook environments (kernels) unpacked with Docker (and reprozip-jupyter is installed)
* Add `--docker-cmd` to reprounzip-docker to select the Docker command (for example `--docker-cmd="sudo docker"`)
* Implement `--expose-port` option for Vagrant and Docker (no need for `--docker-option=-p...`)
* Add docker-machine support to GUI (select which machine to use)
* Better binaries for MacOS
* Automatically register reprounzip-qt to open .RPZ files on Linux
* Register ReproUnzip with Windows from installer
* Add icon from @heng2j

1.0.9 (2017-01-10)
------------------

Bugfixes:
* Fix CentOS Docker image versions
* Remove Fedora Docker images, they don't have tar
* Do include .pyc files in packages, so reproduction take same code path
* Don't use the experiment's resolv.conf file
* Fix handling of files opened in read-write mode

1.0.8 (2016-10-07)
------------------

Behavior changes:

reprounzip/reprounzip/reprounzip.py
# Copyright (C) 2014 New York University
# This file is part of ReproZip which is released under the Revised BSD License
# See file LICENSE for full license details.

from __future__ import unicode_literals

import argparse
import codecs
import locale
import logging
from pkg_resources import iter_entry_points
import sys


__version__ = '0.3.2'


def main():
    """Entry point when called on the command line.
    """
    # Locale
    locale.setlocale(locale.LC_ALL, '')

    # Encoding for output streams
    if str == bytes:
        writer = codecs.getwriter(locale.getpreferredencoding())
        sys.stdout = writer(sys.stdout)
        sys.stderr = writer(sys.stderr)

    # http://bugs.python.org/issue13676
    # This prevents reprozip from reading argv and envp arrays from trace
    if sys.version_info < (2, 7, 3):
        sys.stderr.write("Error: your version of Python, %s, is not "
                         "supported\nVersions before 2.7.3 are affected by "
                         "bug 13676 and will not work with ReproZip\n" %
                         sys.version.split(' ', 1)[0])
        sys.exit(1)

    # Parses command-line

    # General options
    options = argparse.ArgumentParser(add_help=False)
    options.add_argument('-v', '--verbose', action='count', default=1,
                         dest='verbosity',
                         help="augments verbosity level")

    parser = argparse.ArgumentParser(
            description="Reproducible experiments tool.",
            epilog="Please report issues to reprozip-users@vgc.poly.edu",
            parents=[options])
    subparsers = parser.add_subparsers(title="formats", metavar='')

    # Loads commands from plugins
    for entry_point in iter_entry_points('reprounzip.unpackers'):
        setup_function = entry_point.load()
        setup_function(subparsers=subparsers, general_options=options)

    args = parser.parse_args()
    levels = [logging.CRITICAL, logging.WARNING, logging.INFO, logging.DEBUG]
    logging.basicConfig(level=levels[min(args.verbosity, 3)])
    args.func(args)
    sys.exit(0)


if __name__ == '__main__':
    main()
